service-search-nls/src/main/java/fi/nls/oskari/search/util/ELFGeoLocatorCountries.java

```diff
@@ -31,53 +30,118 @@
         if (serviceURL == null) {
             log.warn("ServiceURL not configured. Add property with key", ELFGeoLocatorSearchChannel.PROPERTY_SERVICE_URL);
         }
+        loadCountryMap();
     }
 
-    public Map<String, String> getCountryMap() throws IOException {
-        // not properly configured
-        if (serviceURL == null) {
-            return countries;
+    public void loadCountryMap () {
+        try {
+            // not properly configured
+            if (serviceURL == null) {
+                return ;
+            }
+
+            // synced recently
+            if(System.currentTimeMillis() < lastSync + REFRESH_INTERVAL) {
+                return ;
+            }
+            parseCountryMap(getCountryData());
+            if(countryMap.isEmpty()) {
+                log.debug("Could not get countries");
+            }
         }
-        // synced recently
-        if(System.currentTimeMillis() < lastSync + REFRESH_INTERVAL) {
-            return countries;
+        catch (Exception e) {
+            log.debug(e, "Could not get countries");
         }
+        lastSync = System.currentTimeMillis();
+    }
 
-        // Start from scratch
+    public String getCountryData() throws IOException {
         String countriesUrl = serviceURL + DEFAULT_GETCOUNTRIES_TEMPLATE;
-        StringBuffer buf = new StringBuffer(countriesUrl);
-        log.debug("Server request: " + buf.toString());
-        HttpURLConnection conn = channel.getConnection(buf.toString());
+        log.debug("Server request: " + countriesUrl);
+        HttpURLConnection conn = IOHelper.getConnection(countriesUrl);
         String response = IOHelper.readString(conn);
-        log.debug("Server response: " + response);
+        return response;
+    }
+
+    public void parseCountryMap(String response) {
+        if (response == null || response.isEmpty()) {
+            return ;
+        }
 
         try {
             DocumentBuilder dBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
             Document doc = dBuilder.parse(new ByteArrayInputStream(response.getBytes("UTF-8")));
 
             NodeList nList = doc.getElementsByTagName("Country");
-            countries.clear();
+            countryMap.clear();
             for (int temp = 0; temp < nList.getLength(); temp++) {
                 Node nNode = nList.item(temp);
                 if (nNode.getNodeType() != Node.ELEMENT_NODE) {
                     continue;
                 }
                 Element eElement = (Element) nNode;
-                countries.put(eElement.getElementsByTagName("code").item(0).getTextContent(),
-                        eElement.getElementsByTagName("administrator").item(0).getTextContent());
+                NodeList elements = eElement.getElementsByTagName("administrator");
+                for(int i=0; i<elements.getLength(); i++) {
+                    countryMap.put(eElement.getElementsByTagName("administrator").item(i).getTextContent(),
```

Comment:
Dont call getElementByTagName() inside the loop, you already have the NodeList - call elements.item(i)
Same thing with the next line, grab the value of "code".item(0).getTextContent() before entering the loop.